File: kotlin/Algorithm/src/main/kotlin/Custom002.kt
import java.util.*

data class Node(
    var value: Int = 0,
    var outNode: Node? = null,
    var enterNodeCount: Int = 0
)

fun solution(n: Int, k: Int, connections: List<Pair<String, String>>): Int {
    val nodes: MutableMap<String, Node> = mutableMapOf()

    connections.forEach { (parent, child) ->

        nodes.getOrPut(parent) { Node() }.apply { enterNodeCount += 1 }
        nodes.getOrPut(child) { Node() }.outNode = nodes[parent]
    }

    val queue = LinkedList<Node>().apply {
        nodes.values.filter { node ->
            node.enterNodeCount == 0
        }.forEach { startNode -> add(startNode.apply { value = 1 }) }
    }

    var totalWeight = 0

    while (queue.isNotEmpty()) {
        val currentNode: Node = queue.poll()

        totalWeight += currentNode.value

        currentNode.outNode?.let { outNode ->
            outNode.enterNodeCount--
            outNode.value += currentNode.value

            if (outNode.enterNodeCount == 0) queue.add(outNode)
        }
    }


    return k / totalWeight
}

fun main() {
    for (testNum in 1..readln().toInt()) {
        val inputData: List<Int> = readln().split(" ").map { e -> e.toInt() }
        val n = inputData[0]
        val k = inputData[1]

        val connections = mutableListOf<Pair<String, String>>()
        repeat(n) {
            val connectionData: List<String> = readln().split(" ")
            connections.add(Pair(connectionData[0], connectionData[1]))
        }

        println("#$testNum: ${solution(n, k, connections)}")
    }
}

File: python/Algorithm/CUSTOM002.py
from collections import deque


class Node:
    def __init__(self):
        self.value = 0
        self.out_node = None
        self.enter_node_count = 0


T = int(input())

for test_num in range(T):
    N, K = list(map(int, input().split()))
    nodes = {}

    for _ in range(N):
        parent, child = input().split()

        if parent not in nodes:
            nodes[parent] = Node()

        if child not in nodes:
            nodes[child] = Node()

        nodes[parent].enter_node_count += 1
        nodes[child].out_node = nodes[parent]

    queue = deque([])

    for node_name, node in nodes.items():
        if node.enter_node_count == 0:
            node.value = 1
            queue.append(node)

    total_weight = 0

    while queue:
        current_node = queue.popleft()

        total_weight += current_node.value

        if current_node.out_node:
            current_node.out_node.value += current_node.value
            current_node.out_node.enter_node_count -= 1

            if current_node.out_node.enter_node_count == 0:
                queue.append(current_node.out_node)

    print('#%d: %d' % (test_num + 1, K // total_weight))
